primitives/blockchain/src/error.rs

```diff
@@ -17,142 +17,143 @@
-	CallResultDecode(&'static str, CodecError),
+	#[error("Error decoding call result of {0}")]
+	CallResultDecode(&'static str, #[source] CodecError),
 	/// Error converting a parameter between runtime and node.
-	#[display(fmt = "Error converting `{}` between runtime and node", _0)]
-	#[from(ignore)]
+	#[error("Error converting `{0}` between runtime and node")]
 	RuntimeParamConversion(String),
 	/// Changes tries are not supported.
-	#[display(fmt = "Changes tries are not supported by the runtime")]
+	#[error("Changes tries are not supported by the runtime")]
 	ChangesTriesNotSupported,
 	/// Error reading changes tries configuration.
-	#[display(fmt = "Error reading changes tries configuration")]
+	#[error("Error reading changes tries configuration")]
 	ErrorReadingChangesTriesConfig,
 	/// Key changes query has failed.
-	#[display(fmt = "Failed to check changes proof: {}", _0)]
-	#[from(ignore)]
+	#[error("Failed to check changes proof: {0}")]
 	ChangesTrieAccessFailed(String),
 	/// Last finalized block not parent of current.
-	#[display(fmt = "Did not finalize blocks in sequential order.")]
-	#[from(ignore)]
+	#[error("Did not finalize blocks in sequential order.")]
 	NonSequentialFinalization(String),
 	/// Safety violation: new best block not descendent of last finalized.
-	#[display(fmt = "Potential long-range attack: block not in finalized chain.")]
+	#[error("Potential long-range attack: block not in finalized chain.")]
 	NotInFinalizedChain,
 	/// Hash that is required for building CHT is missing.
-	#[display(fmt = "Failed to get hash of block for building CHT")]
+	#[error("Failed to get hash of block for building CHT")]
 	MissingHashRequiredForCHT,
 	/// Invalid calculated state root on block import.
-	#[display(fmt = "Calculated state root does not match.")]
+	#[error("Calculated state root does not match.")]
 	InvalidStateRoot,
 	/// Incomplete block import pipeline.
-	#[display(fmt = "Incomplete block import pipeline.")]
+	#[error("Incomplete block import pipeline.")]
 	IncompletePipeline,
-	#[display(fmt = "Transaction pool not ready for block production.")]
+	#[error("Transaction pool not ready for block production.")]
 	TransactionPoolNotReady,
-	#[display(fmt = "Database: {}", _0)]
-	DatabaseError(sp_database::error::DatabaseError),
+	#[error("Database")]
+	DatabaseError(#[from] sp_database::error::DatabaseError),
 	/// A convenience variant for String
-	#[display(fmt = "{}", _0)]
+	#[error("{0}")]
 	Msg(String),
 }
 
-impl error::Error for Error {
-	fn source(&self) -> Option<&(dyn error::Error + 'static)> {
-		match self {
-			Error::Consensus(e) => Some(e),
-			Error::Blockchain(e) => Some(e),
-			_ => None,
-		}
-	}
-}
-
 impl<'a> From<&'a str> for Error {
 	fn from(s: &'a str) -> Self {
 		Error::Msg(s.into())
 	}
 }
 
+impl From<String> for Error {
+	fn from(s: String) -> Self {
+		Error::Msg(s)
+	}
+}
+
+
+impl From<Box<dyn sp_state_machine::Error + Send + Sync>> for Error {
```

Comment:
Hmm. The error is already `Send + Sync`. Is this still generating a "different" type from the POV of the compiler?

Reply:
I thought so too, but both variants (with explicit `Send + Sync`) are required unfortunately to make the compiler happy.

Reply:
Filed and upstream issue https://github.com/rust-lang/rust/issues/78700